### ISSUE TITLE
windows issues, dbgeng adapter returned -1 or -3

### ISSUE THREAD

**Issue by lwerdna** [Vector35/deprecated_python_debugger#113]
trace from user Aaron on slack:

```
Traceback (most recent call last):
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\dockwidgets\ControlsWidget.py", line 234, in perform_run_thread
    self.debug_state.run()
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\binjaplug.py", line 593, in run
    self.exec()
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\binjaplug.py", line 640, in exec
    raise e
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\binjaplug.py", line 635, in exec
    self.adapter.exec(fpath, self.command_line_args, terminal=self.request_terminal_emulator)
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 116, in exec
    return self.submit(lambda: self.adapter.exec(path, args, **kwargs))
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 84, in submit
    raise result
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 48, in worker
    self.results[index] = (True, job())
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 116, in <lambda>
    return self.submit(lambda: self.adapter.exec(path, args, **kwargs))
  File "C:\Users\Me\AppData\Roaming\Binary Ninja\plugins\debugger\dbgeng.py", line 192, in exec
    raise Exception('unable to launch "%s", dbgeng adapter returned %d' % (cmdline, rc))
Exception: unable to launch "C:\myexe.exe", dbgeng adapter returned -1
```

trace from VM, this one with -3 returned:

```
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\dockwidgets\ControlsWidget.py", line 234, in perform_run_thread
    self.debug_state.run()
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\binjaplug.py", line 593, in run
    self.exec()
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\binjaplug.py", line 640, in exec
    raise e
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\binjaplug.py", line 635, in exec
    self.adapter.exec(fpath, self.command_line_args, terminal=self.request_terminal_emulator)
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 116, in exec
    return self.submit(lambda: self.adapter.exec(path, args, **kwargs))
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 84, in submit
    raise result
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 48, in worker
    self.results[index] = (True, job())
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\QueuedAdapter.py", line 116, in <lambda>
    return self.submit(lambda: self.adapter.exec(path, args, **kwargs))
  File "C:\Users\Andrew Lamer\AppData\Roaming\Binary Ninja\plugins\debugger\dbgeng.py", line 193, in exec
    raise Exception('unable to launch "%s", dbgeng adapter returned %d' % (cmdline, rc))
Exception: unable to launch "'C:\Users\Andrew Lamer\repos\vector35\debugger\testbins\helloworld_func_x64-windows.exe'", dbgeng adapter returned -3
```

**Comment by lwerdna** [comment]
The dbgeng client will reject a single-quoted path to target which contains spaces.

Function shlex.quote() returns single-quoted paths. Not just paths, arguments too, compare:

```
c:\Users\Andrew Lamer\repos\vector35\debugger\testbins>helloworld_x64-windows.exe 'one two' three
Hello, world!
argc: 4
argv[0]: helloworld_x64-windows.exe
argv[1]: 'one
argv[2]: two'
argv[3]: three
```

```
c:\Users\Andrew Lamer\repos\vector35\debugger\testbins>helloworld_x64-windows.exe "one two" three
Hello, world!
argc: 3
argv[0]: helloworld_x64-windows.exe
argv[1]: one two
argv[2]: three
```

I'm tempted to declare `shlex.quote()` unusable for paths and arguments prior to being sent to `CreateProcess()` but it's more like I'm getting something wrong than python. For now, double-quoting manually. This should solve the -3 issue.

**Comment by lwerdna** [comment]
MSDN is not very clear on exactly what starts and ends a session: https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugging-session-and-execution-model

The debug client has an `EndSession()` method but not a corresponding `StartSession()`, see https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/dbgeng/nf-dbgeng-idebugclient-endsession

You must know, because whatever thread calls WaitForEvent() must be the same thread that started the session. Our QueuedAdapter is not only for serialization of commands to the debugger, it's for ensuring a "same thread" policy when interacting with dbgeng.

I thought that creating the client interface was NOT starting a session, but calling `CreateProcess()` and waiting for `DEBUG_SESSION_ACTIVE` was. But this is not the case, it appears that just getting a client interface with `DebugCreate()` is starting a session, and my source is how `WaitForEvent()` behaves (hanging or not). So now I removed all `setup()` and `teardown()` idea from DebugAdapter. It was there so separate initialization code from DLL_ATTACH. Now the dbgeng idea of session will run in parallel with the casual idea of debugger session. Everytime you exec a target, we reacquire COM interfaces, and everytime you end a target, we release COM interfaces. It doesn't feel elegant, but there's no cost I've noticed.
